lib/sum_of_balances_latest/main.py

```diff
@@ -1,74 +1,95 @@
     """
-    institutions = client.institutions.get(user_uuid)
 
-    # Sum the most recent balances from each account within our time window.
+    # Sum the most recent balances from each account.
     sum_of_balances_latest = 0
-    for institution in institutions:
-        institution_id = institution.institution_id
-        balances = client.balances.get(
-            user_uuid,
-            institution_id,
+
+    institutions = await api_client.institutions.get(user_uuid=user_uuid)
+    utc_starttime = utc_time - cutoff_interval
+
+    inst_coroutines = [
+        api_client.balances.get(
+            user_uuid=user_uuid,
+            institution_id=institution.institution_id,
             utc_starttime=utc_starttime,
-            utc_endtime=utc_endtime,
+            utc_endtime=utc_time,
         )
+        for institution in institutions
+    ]
+
+    r = await asyncio.gather(*inst_coroutines)
+    institution_ids = [institution.institution_id for institution in institutions]
+
+    balances_by_institution_id = dict(zip(institution_ids, r))
 
-        # Add latest account balance for each account within the institution.
-        account_numbers = {balance.account_number for balance in balances}
-        for account_number in account_numbers:
-            depository_balances = [
-                balance
-                for balance in balances
-                if balance.account_type == "depository"
-                and balance.account_number == account_number
-            ]
-            if depository_balances:
-                # Find latest balance record by sorting in time.
-                latest_balance = sorted(
-                    depository_balances,
-                    key=lambda balance: balance.ts,
-                    reverse=True,
-                )[0]
-                sum_of_balances_latest += latest_balance.balance
+    depository_balance_df_by_institution_id: Dict[str, pd.DataFrame] = {}
+    for institution_id, balances in balances_by_institution_id.items():
+
+        depository_balances = [
+            balance for balance in balances if balance.account_type == "depository"
+        ]
+
+        if depository_balances:
+            depository_balance_df_by_institution_id[institution_id] = pd.DataFrame(
+                [balance.dict() for balance in depository_balances]
+            )
+
+    # Add latest account balance for each account within the institution.
+    for institution_id, balances_df in depository_balance_df_by_institution_id.items():
+        unique_account_numbers = set(balances_df["account_number"].to_list())
+        for account_number in unique_account_numbers:
+            filter_account_number = balances_df.account_number == account_number
+
+            sorted_df_by_timestamp = balances_df[filter_account_number].sort_values(
+                by="ts", ascending=False
+            )
+            sum_of_balances_latest += sorted_df_by_timestamp.iloc[0]["balance"]
```

Comment:
note for future refactor.
since we're using pandas, I think all of the above can be achieved in one line.
something like df.group_by('account_number').latest().sum() or something like that.
AOK as it is for now tho